* Shabarinath Chandran
* EAI6000
* 02/24/2020

# Titanic Survival Prediction

# Importing Necessary Libraries and data set


In [148]:
import numpy as np
import pandas as pd

In [157]:
train = pd.read_csv("/Users/shivrajchandran/Downloads/titanic/train.csv")
test = pd.read_csv("/Users/shivrajchandran/Downloads/titanic/test.csv")
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [158]:
print(train.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [159]:
train.sample(5)

PassengerId  Survived  Pclass                                    Name  \
558          559         1       1  Taussig, Mrs. Emil (Tillie Mandelbaum)   
368          369         1       3                     Jermyn, Miss. Annie   
232          233         0       2               Sjostedt, Mr. Ernst Adolf   
175          176         0       3                  Klasen, Mr. Klas Albin   
154          155         0       3                   Olsen, Mr. Ole Martin   

        Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
558  female  39.0      1      1     110413  79.6500   E67        S  
368  female   NaN      0      0      14313   7.7500   NaN        Q  
232    male  59.0      0      0     237442  13.5000   NaN        S  
175    male  18.0      1      1     350404   7.8542   NaN        S  
154    male   NaN      0      0  Fa 265302   7.3125   NaN        S

In [160]:
train.describe(include = "all")

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Lefebre, Miss. Mathilde  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch    Ticket        Fare Cabin  \
count   714.000000  891.000000  891.000000       891  891.000000   204   
unique         NaN         NaN         NaN       681         NaN   147   
top            NaN         NaN         NaN  CA. 2343         NaN    G6   
freq           NaN         NaN         NaN         7         NaN     4   
mean     29.699118    0.523008    0.381594       NaN   32.204208   NaN   
std      14.526497    1.102743    0.806057       NaN   49.693429   NaN   
min       0.420000    0.000000    0.000000       NaN    0.000000   NaN   
25%      20.125000    0.000000    0.000000       NaN    7.910400   NaN   
50%      28.000000    0.000000    0.000000       NaN   14.454200   NaN   
75%      38.000000    1.000000    0.000000       NaN   31.000000   NaN   
max      80.000000    8.000000    6.000000       NaN  512.329200   NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

# Findings from data

* There are a total of 891 passengers in the training set. However, the Age feature is missing approximately 19.8% of its values.
* Age is not available for few, but as the Age feature is important to survival, we will fill these gaps.
* It is assumed that females are more likely to survive.
* Also, Young children and People of higher socio economic are more likely to survive.

In [161]:
print(pd.isnull(train).sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


* We can see that except for the above mentioned missing values, no NaN values exist.


* Now,Sorting the ages into logical categories as we can predict the survival rate based on the categories of age group. Babies are most likely to survive than other groups.

In [162]:
train["Age"] = train["Age"].fillna(-0.5)
test["Age"] = test["Age"].fillna(-0.5)
bins = [-1, 0, 5, 12, 18, 24, 35, 60, np.inf]
labels = ['Unknown', 'Baby', 'Child', 'Teenager', 'Student', 'Young Adult', 'Adult', 'Senior']
train['AgeGroup'] = pd.cut(train["Age"], bins, labels = labels)
test['AgeGroup'] = pd.cut(test["Age"], bins, labels = labels)

## Cleaning the Data


In [164]:
test.describe(include="all")

PassengerId      Pclass                Name   Sex         Age  \
count    418.000000  418.000000                 418   418  418.000000   
unique          NaN         NaN                 418     2         NaN   
top             NaN         NaN  Herman, Miss. Kate  male         NaN   
freq            NaN         NaN                   1   266         NaN   
mean    1100.500000    2.265550                 NaN   NaN   23.941388   
std      120.810458    0.841838                 NaN   NaN   17.741080   
min      892.000000    1.000000                 NaN   NaN   -0.500000   
25%      996.250000    1.000000                 NaN   NaN    9.000000   
50%     1100.500000    3.000000                 NaN   NaN   24.000000   
75%     1204.750000    3.000000                 NaN   NaN   35.750000   
max     1309.000000    3.000000                 NaN   NaN   76.000000   

             SibSp       Parch    Ticket        Fare            Cabin  \
count   418.000000  418.000000       418  417.000000               91   
unique         NaN         NaN       363         NaN               76   
top            NaN         NaN  PC 17608         NaN  B57 B59 B63 B66   
freq           NaN         NaN         5         NaN                3   
mean      0.447368    0.392344       NaN   35.627188              NaN   
std       0.896760    0.981429       NaN   55.907576              NaN   
min       0.000000    0.000000       NaN    0.000000              NaN   
25%       0.000000    0.000000       NaN    7.895800              NaN   
50%       0.000000    0.000000       NaN   14.454200              NaN   
75%       1.000000    0.000000       NaN   31.500000              NaN   
max       8.000000    9.000000       NaN  512.329200              NaN   

       Embarked     AgeGroup   CabinBool  
count       418          418  418.000000  
unique        3            8         NaN  
top           S  Young Adult         NaN  
freq        270           96         NaN  
mean        NaN          NaN    0.217703  
std         NaN          NaN    0.413179  
min         NaN          NaN    0.000000  
25%         NaN          NaN    0.000000  
50%         NaN          NaN    0.000000  
75%         NaN          NaN    0.000000  
max         NaN          NaN    1.000000

* We have a total of 418 passengers.
* 1 value from the Fare feature is missing.
* Around 20.5% of the Age feature is missing, we will need to fill that in.

* Dropping the Cabin feature since not a lot more useful information can be extracted from it.

In [165]:
train = train.drop(['Cabin'], axis = 1)
test = test.drop(['Cabin'], axis = 1)

* Dropping the Ticket feature since it's unlikely to yield any useful information

In [166]:

train = train.drop(['Ticket'], axis = 1)
test = test.drop(['Ticket'], axis = 1)

* Filling in the missing values in the Embarked feature

In [167]:

print("Number of people embarking in Southampton (S):")
southampton = train[train["Embarked"] == "S"].shape[0]
print(southampton)

print("Number of people embarking in Cherbourg (C):")
cherbourg = train[train["Embarked"] == "C"].shape[0]
print(cherbourg)

print("Number of people embarking in Queenstown (Q):")
queenstown = train[train["Embarked"] == "Q"].shape[0]
print(queenstown)

Number of people embarking in Southampton (S):
644
Number of people embarking in Cherbourg (C):
168
Number of people embarking in Queenstown (Q):
77


* As it's clear that the majority of people embarked in Southampton (S). We are filling in the missing values with S.

In [168]:
train = train.fillna({"Embarked": "S"}) # Replacing the missing values in the Embarked feature with S

* Filling in the missing values in the Age feature. Since a higher percentage of values are missing, it would be illogical to fill all of them with the same value.  

In [169]:
combine = [train, test] #creating a combined group of both datasets

for dataset in combine: #extract a title for each Name in the train and test datasets
    dataset['Title'] = dataset.Name.str.extract(' ([A-Za-z]+)\.', expand=False)

pd.crosstab(train['Title'], train['Sex'])

Sex       female  male
Title                 
Capt           0     1
Col            0     2
Countess       1     0
Don            0     1
Dr             1     6
Jonkheer       0     1
Lady           1     0
Major          0     2
Master         0    40
Miss         182     0
Mlle           2     0
Mme            1     0
Mr             0   517
Mrs          125     0
Ms             1     0
Rev            0     6
Sir            0     1

In [170]:
#replace various titles with more common names
for dataset in combine:
    dataset['Title'] = dataset['Title'].replace(['Lady', 'Capt', 'Col',
    'Don', 'Dr', 'Major', 'Rev', 'Jonkheer', 'Dona'], 'Rare')
    
    dataset['Title'] = dataset['Title'].replace(['Countess', 'Lady', 'Sir'], 'Royal')
    dataset['Title'] = dataset['Title'].replace('Mlle', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Ms', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Mme', 'Mrs')

train[['Title', 'Survived']].groupby(['Title'], as_index=False).mean()

Title  Survived
0  Master  0.575000
1    Miss  0.702703
2      Mr  0.156673
3     Mrs  0.793651
4    Rare  0.285714
5   Royal  1.000000

In [171]:
#map each of the title groups to a numerical value
title_mapping = {"Mr": 1, "Miss": 2, "Mrs": 3, "Master": 4, "Royal": 5, "Rare": 6}
for dataset in combine:
    dataset['Title'] = dataset['Title'].map(title_mapping)
    dataset['Title'] = dataset['Title'].fillna(0)

train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch     Fare Embarked     AgeGroup  CabinBool  Title  
0      0   7.2500        S      Student          0      1  
1      0  71.2833        C        Adult          1      3  
2      0   7.9250        S  Young Adult          0      2  
3      0  53.1000        S  Young Adult          1      3  
4      0   8.0500        S  Young Adult          0      1

* Trying to predict the missing Age values from the most common age for their Title.

In [172]:
# fill missing age with mode age group for each title
mr_age = train[train["Title"] == 1]["AgeGroup"].mode() #Young Adult
miss_age = train[train["Title"] == 2]["AgeGroup"].mode() #Student
mrs_age = train[train["Title"] == 3]["AgeGroup"].mode() #Adult
master_age = train[train["Title"] == 4]["AgeGroup"].mode() #Baby
royal_age = train[train["Title"] == 5]["AgeGroup"].mode() #Adult
rare_age = train[train["Title"] == 6]["AgeGroup"].mode() #Adult

age_title_mapping = {1: "Young Adult", 2: "Student", 3: "Adult", 4: "Baby", 5: "Adult", 6: "Adult"}


for x in range(len(train["AgeGroup"])):
    if train["AgeGroup"][x] == "Unknown":
        train["AgeGroup"][x] = age_title_mapping[train["Title"][x]]
        
for x in range(len(test["AgeGroup"])):
    if test["AgeGroup"][x] == "Unknown":
        test["AgeGroup"][x] = age_title_mapping[test["Title"][x]]

Now that we've filled in the missing values with little accuracy, we can map each age group to a numerical value.

In [173]:
#map each Age value to a numerical value
age_mapping = {'Baby': 1, 'Child': 2, 'Teenager': 3, 'Student': 4, 'Young Adult': 5, 'Adult': 6, 'Senior': 7}
train['AgeGroup'] = train['AgeGroup'].map(age_mapping)
test['AgeGroup'] = test['AgeGroup'].map(age_mapping)

train.head()

#dropping the Age feature for now, might change
train = train.drop(['Age'], axis = 1)
test = test.drop(['Age'], axis = 1)

* Dropping the name feature since it contains no more useful information.

In [174]:
train = train.drop(['Name'], axis = 1)
test = test.drop(['Name'], axis = 1)

* Mapping each Sex value to a numerical value

In [176]:
sex_mapping = {"male": 0, "female": 1}
train['Sex'] = train['Sex'].map(sex_mapping)
test['Sex'] = test['Sex'].map(sex_mapping)

train.head()

PassengerId  Survived  Pclass  Sex  SibSp  Parch     Fare Embarked  \
0            1         0       3  NaN      1      0   7.2500        S   
1            2         1       1  NaN      1      0  71.2833        C   
2            3         1       3  NaN      0      0   7.9250        S   
3            4         1       1  NaN      1      0  53.1000        S   
4            5         0       3  NaN      0      0   8.0500        S   

   AgeGroup  CabinBool  Title  
0       4.0          0      1  
1       6.0          1      3  
2       5.0          0      2  
3       5.0          1      3  
4       5.0          0      1

* Mapping each Embarked value to a numerical value

In [177]:
embarked_mapping = {"S": 1, "C": 2, "Q": 3}
train['Embarked'] = train['Embarked'].map(embarked_mapping)
test['Embarked'] = test['Embarked'].map(embarked_mapping)

train.head()

PassengerId  Survived  Pclass  Sex  SibSp  Parch     Fare  Embarked  \
0            1         0       3  NaN      1      0   7.2500         1   
1            2         1       1  NaN      1      0  71.2833         2   
2            3         1       3  NaN      0      0   7.9250         1   
3            4         1       1  NaN      1      0  53.1000         1   
4            5         0       3  NaN      0      0   8.0500         1   

   AgeGroup  CabinBool  Title  
0       4.0          0      1  
1       6.0          1      3  
2       5.0          0      2  
3       5.0          1      3  
4       5.0          0      1

* Separating the fare values into some logical groups as well as filling in the single missing value in the test dataset.

In [178]:
#filling in missing Fare value in test set based on mean fare for that Pclass 
for x in range(len(test["Fare"])):
    if pd.isnull(test["Fare"][x]):
        pclass = test["Pclass"][x] #Pclass = 3
        test["Fare"][x] = round(train[train["Pclass"] == pclass]["Fare"].mean(), 4)
        
#mapping Fare values into groups of numerical values
train['FareBand'] = pd.qcut(train['Fare'], 4, labels = [1, 2, 3, 4])
test['FareBand'] = pd.qcut(test['Fare'], 4, labels = [1, 2, 3, 4])

#dropping Fare values
train = train.drop(['Fare'], axis = 1)
test = test.drop(['Fare'], axis = 1)

In [179]:
train.head()

PassengerId  Survived  Pclass  Sex  SibSp  Parch  Embarked  AgeGroup  \
0            1         0       3  NaN      1      0         1       4.0   
1            2         1       1  NaN      1      0         2       6.0   
2            3         1       3  NaN      0      0         1       5.0   
3            4         1       1  NaN      1      0         1       5.0   
4            5         0       3  NaN      0      0         1       5.0   

   CabinBool  Title FareBand  
0          0      1        1  
1          1      3        4  
2          0      2        2  
3          1      3        4  
4          0      1        2

In [180]:
test.head()

PassengerId  Pclass  Sex  SibSp  Parch  Embarked  AgeGroup  CabinBool  \
0          892       3  NaN      0      0         3       5.0          0   
1          893       3  NaN      1      0         1       6.0          0   
2          894       2  NaN      0      0         3       7.0          0   
3          895       3  NaN      0      0         1       5.0          0   
4          896       3  NaN      1      1         1       4.0          0   

   Title FareBand  
0      1        1  
1      3        1  
2      1        2  
3      1        2  
4      3        2

## PREDICTION

### Splitting the Training Data
Using part of our training data to test the accuracy of our different models. 22% in this case.

In [136]:
from sklearn.model_selection import train_test_split

predictors = train.drop(['Survived', 'PassengerId'], axis=1)
target = train["Survived"]
x_train, x_val, y_train, y_val = train_test_split(predictors, target, test_size = 0.22, random_state = 0)

### Testing Different Models


For each model, we set the model, fit it with 80% of our training data, predict for 20% of the training data and check the accuracy.

In [182]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()
logreg.fit(x_train, y_train)
y_pred = logreg.predict(x_val)
acc_logreg = round(accuracy_score(y_pred, y_val) * 100, 2)
print(acc_logreg)

79.19


In [183]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier

decisiontree = DecisionTreeClassifier()
decisiontree.fit(x_train, y_train)
y_pred = decisiontree.predict(x_val)
acc_decisiontree = round(accuracy_score(y_pred, y_val) * 100, 2)
print(acc_decisiontree)

79.19


In [184]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier

randomforest = RandomForestClassifier()
randomforest.fit(x_train, y_train)
y_pred = randomforest.predict(x_val)
acc_randomforest = round(accuracy_score(y_pred, y_val) * 100, 2)
print(acc_randomforest)

80.2


In [185]:
# Gradient Boosting Classifier
from sklearn.ensemble import GradientBoostingClassifier

gbk = GradientBoostingClassifier()
gbk.fit(x_train, y_train)
y_pred = gbk.predict(x_val)
acc_gbk = round(accuracy_score(y_pred, y_val) * 100, 2)
print(acc_gbk)

84.77


* Based on the predictions from the above models, we can see that Gradient Boosting Classifier is the best model as it has an accuracy close to 85%.

* References
* https://www.kaggle.com/startupsci/titanic-data-science-solutions
* https://www.kaggle.com/jeffd23/scikit-learn-ml-from-start-to-finish?scriptVersionId=320209
